README.md
# Halloumi Cloudwatch Dashboard

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg)](code_of_conduct.md)
[![GitHub](https://img.shields.io/badge/github-sentiampc%2Fhalloumi--cloudwatch--dashboard-orange.svg)](https://github.com/sentiampc/halloumi-cloudwatch-dashboard)
[![PyPI version fury.io](https://badge.fury.io/py/halloumi-cloudwatch-dashboard.svg)](https://pypi.python.org/pypi/halloumi-cloudwatch-dashboard/)
[![npm version](https://badge.fury.io/js/halloumi-cloudwatch-dashboard.svg)](https://badge.fury.io/js/halloumi-cloudwatch-dashboard)
[![GitHub license](https://img.shields.io/github/license/sentiampc/halloumi-cloudwatch-dashboard.svg)](https://github.com/sentiampc/halloumi-cloudwatch-dashboard/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/sentiampc/halloumi-cloudwatch-dashboard.svg)](https://GitHub.com/sentiampc/halloumi-cloudwatch-dashboard/releases/)

Create a dashboard in the AWS Cloudwatch using the best practices from Halloumi.

## Install

From pip:
```bash
pip install halloumi-cloudwatch-dashboard
```

From npm:
```bash
npm require halloumi-cloudwatch-dashboard
```

## API Documentation

Check [API Documentation](./API.md)

## Usage

### Python

```python
from aws_cdk import core
from halloumi-cloudwatch-dashboard import Dashboard

app = core.App()

stack = core.Stack(app, 'MainStack')
...

Dashboard(
    stack,
    'Dashboard',
    dashboardName='MyDashboard'
)
```

### Typescript

```typescript
import * as cdk from '@aws-cdk/core';
import { Dashboard } from '@halloumi-cloudwatch-dashboard';

export class CdkWorkshopStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    Dashboard(this, 'Dashboard', dashboardName='MyDashboard');
  }
}
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.1 (2021-02-08)


### Features

* First release ([b20c910](https://github.com/sentiampc/halloumi-cloudwatch-dashboard/commit/b20c910e9131d2d1698d7457d1937711d6188d63))
* Initiate project ([c2e657d](https://github.com/sentiampc/halloumi-cloudwatch-dashboard/commit/c2e657d58bb0e6cdb1acd627a779cc1b9534c8b2))


### Bug Fixes

* Removing ECS from title for the ASG ([#6](https://github.com/sentiampc/halloumi-cloudwatch-dashboard/issues/6)) ([a17421a](https://github.com/sentiampc/halloumi-cloudwatch-dashboard/commit/a17421ad69eb51f6e0126ae46967e096fdeb5a1c))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


test/rds.test.ts
import { expect, haveResource } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Dashboard } from '../src';

test('dashboard - RDS', () => {
  const stack = new cdk.Stack();
  let body = {
    'Fn::Join': [
      '',
      [
        '{"start":"-PT24H","periodOverride":"inherit","widgets":[{"type":"text","width":24,"height":1,"x":0,"y":0,"properties":{"markdown":"# hello-world - RDS"}},{"type":"metric","width":12,"height":6,"x":0,"y":1,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"CPUUtilization (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","CPUUtilization","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":7,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"FreeableMemory (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","FreeableMemory","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":13,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"FreeLocalStorage (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","FreeLocalStorage","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":19,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"DatabaseConnections (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","DatabaseConnections","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":25,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"DiskQueueDepth (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","DiskQueueDepth","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":31,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"NetworkReceiveThroughput (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","NetworkReceiveThroughput","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":37,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"NetworkTransmitThroughput (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/RDS","NetworkTransmitThroughput","DBInstanceIdentifier","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}}]}',
      ],
    ],
  };
  new Dashboard(stack, 'Dashboard', { rdsName: 'hello-world' });
  expect(stack).to(haveResource('AWS::CloudWatch::Dashboard', { DashboardBody: body }));
});


test/redis.test.ts
import { expect, haveResource } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Dashboard } from '../src';

test('dashboard - Redis', () => {
  const stack = new cdk.Stack();
  let body = {
    'Fn::Join': [
      '',
      [
        '{"start":"-PT24H","periodOverride":"inherit","widgets":[{"type":"text","width":24,"height":1,"x":0,"y":0,"properties":{"markdown":"# hello-world - Redis"}},{"type":"metric","width":12,"height":6,"x":0,"y":1,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"CPUUtilization (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","CPUUtilization","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":7,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"FreeableMemory (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","FreeableMemory","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":13,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"SwapUsage (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","SwapUsage","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":19,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"NetworkBytesIn (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","NetworkBytesIn","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":25,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"NetworkBytesOut (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","NetworkBytesOut","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":31,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"CurrConnections (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","CurrConnections","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":37,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"Evictions (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","Evictions","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":43,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"Reclaimed (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","Reclaimed","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":49,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"CacheHits (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","CacheHits","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":55,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"CacheMisses (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","CacheMisses","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":61,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"ReplicationBytes (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","ReplicationBytes","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":67,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"ReplicationLag (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","ReplicationLag","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}},{"type":"metric","width":12,"height":6,"x":0,"y":73,"properties":{"view":"timeSeries","region":"',
        {
          Ref: 'AWS::Region',
        },
        '","metrics":[[{"label":"BytesUsedForCache (expected)","color":"#666666","expression":"ANOMALY_DETECTION_BAND(m1, 2)"}],["AWS/ElastiCache","BytesUsedForCache","CacheClusterId","hello-world",{"yAxis":"right","id":"m1"}]],"yAxis":{}}}]}',
      ],
    ],
  };
  new Dashboard(stack, 'Dashboard', { elasticacheName: 'hello-world' });
  expect(stack).to(haveResource('AWS::CloudWatch::Dashboard', { DashboardBody: body }));
});


test/dashboard.test.ts
import { expect, haveResource } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Dashboard } from '../src';

test('dashboard', () => {
  const stack = new cdk.Stack();
  new Dashboard(stack, 'Dashboard');
  expect(stack).to(haveResource('AWS::CloudWatch::Dashboard', { DashboardBody: '{"start":"-PT24H","periodOverride":"inherit","widgets":[]}' }));
});
test('dashboard - name', () => {
  const stack = new cdk.Stack();
  new Dashboard(stack, 'Dashboard', { dashboardName: 'TheName' });
  expect(stack).to(haveResource('AWS::CloudWatch::Dashboard', { DashboardName: 'TheName' }));
});
test('dashboard - start', () => {
  const stack = new cdk.Stack();
  new Dashboard(stack, 'Dashboard', { start: '-PT6H' });
  expect(stack).to(haveResource('AWS::CloudWatch::Dashboard', { DashboardBody: '{"start":"-PT6H","periodOverride":"inherit","widgets":[]}' }));
});

.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');

const project = new AwsCdkConstructLibrary({
  author: 'Sentia MPC',
  authorAddress: 'support.mpc@sentia.com',
  cdkVersion: '1.73.0',
  jsiiFqn: 'projen.AwsCdkConstructLibrary',
  name: 'halloumi-cloudwatch-dashboard',
  repositoryUrl: 'https://github.com/sentiampc/halloumi-cloudwatch-dashboard.git',

  /* AwsCdkConstructLibraryOptions */
  // cdkAssert: true,                                                                   /* Install the @aws-cdk/assert library? */
  cdkDependencies: [
    '@aws-cdk/aws-cloudwatch',
    '@aws-cdk/core',
  ], /* Which AWS CDK modules (those that start with "@aws-cdk/") does this library require when consumed? */
  // cdkTestDependencies: undefined,                                                    /* AWS CDK modules required for testing. */
  // cdkVersionPinning: false,                                                          /* Use pinned version instead of caret version for CDK. */

  /* ConstructLibraryOptions */
  // catalog: undefined,                                                                /* Libraries will be picked up by the construct catalog when they are published to npm as jsii modules and will be published under:. */

  /* JsiiProjectOptions */
  // compat: false,                                                                     /* Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: '.compatignore',                                                     /* Name of the ignore file for API compatibility tests. */
  // docgen: true,                                                                      /* Automatically generate API.md from jsii. */
  // eslint: true,                                                                      /* Install eslint. */
  // eslintOptions: undefined,                                                          /* Eslint options. */
  // publishToGo: undefined,                                                            /* Publish Go bindings to a git repository. */
  // publishToMaven: undefined,                                                         /* Publish to maven. */
  // publishToNuget: undefined,                                                         /* Publish to NuGet. */
  publishToPypi: {
    distName: 'halloumi-cloudwatch-dashboard',
    module: 'halloumi_cloudwatch_dashboard',
  }, /* Publish to pypi. */
  // rootdir: '.',                                                                      /* undefined */

  /* NodePackageOptions */
  // allowLibraryDependencies: true,                                                    /* Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // authorEmail: undefined,                                                            /* Author's e-mail. */
  // authorName: undefined,                                                             /* Author's name. */
  // authorOrganization: undefined,                                                     /* Author's Organization. */
  // authorUrl: undefined,                                                              /* Author's URL / Website. */
  // autoDetectBin: true,                                                               /* Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                                    /* Binary programs vended with your module. */
  // bundledDeps: undefined,                                                            /* List of dependencies to bundle into this module. */
  // deps: [],                                                                          /* Runtime dependencies of this module. */
  // description: undefined,                                                            /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],                                                                       /* Build dependencies for this module. */
  // entrypoint: 'lib/index.js',                                                        /* Module entrypoint (`main` in `package.json`). */
  // homepage: undefined,                                                               /* Package's Homepage / Website. */
  // keywords: undefined,                                                               /* Keywords to include in `package.json`. */
  // license: 'Apache-2.0',                                                             /* License's SPDX identifier. */
  // licensed: true,                                                                    /* Indicates if a license should be added. */
  // maxNodeVersion: undefined,                                                         /* Minimum node.js version to require via `engines` (inclusive). */
  // minNodeVersion: undefined,                                                         /* Minimum Node.js version to require via package.json `engines` (inclusive). */
  // npmAccess: undefined,                                                              /* Access level of the npm package. */
  // npmDistTag: 'latest',                                                              /* Tags can be used to provide an alias instead of version numbers. */
  // npmRegistryUrl: 'https://registry.npmjs.org',                                      /* The base URL of the npm package registry. */
  // npmTaskExecution: NpmTaskExecution.PROJEN,                                         /* Determines how tasks are executed when invoked as npm scripts (yarn/npm run xyz). */
  // packageManager: NodePackageManager.YARN,                                           /* The Node Package Manager used to execute scripts. */
  // packageName: undefined,                                                            /* The "name" in package.json. */
  // peerDependencyOptions: undefined,                                                  /* Options for `peerDeps`. */
  // peerDeps: [],                                                                      /* Peer dependencies for this module. */
  // projenCommand: 'npx projen',                                                       /* The shell command to use in order to run the projen CLI. */
  // repository: undefined,                                                             /* The repository is the location where the actual code for your package lives. */
  // repositoryDirectory: undefined,                                                    /* If the package.json for your package is not in the root directory (for example if it is part of a monorepo), you can specify the directory in which it lives. */
  // scripts: {},                                                                       /* npm scripts to include. */
  // stability: undefined,                                                              /* Package's Stability. */

  /* NodeProjectOptions */
  antitamper: true, /* Checks that after build there are no modified files on git. */
  // buildWorkflow: undefined,                                                          /* Define a GitHub workflow for building PRs. */
  // codeCov: false,                                                                    /* Define a GitHub workflow step for sending code coverage metrics to https://codecov.io/ Uses codecov/codecov-action@v1 A secret is required for private repos. Configured with @codeCovTokenSecret. */
  // codeCovTokenSecret: undefined,                                                     /* Define the secret name for a specified https://codecov.io/ token A secret is required to send coverage for private repositories. */
  // copyrightOwner: undefined,                                                         /* License copyright owner. */
  // copyrightPeriod: undefined,                                                        /* The copyright years to put in the LICENSE file. */
  defaultReleaseBranch: 'master', /* The name of the main release branch. */
  // dependabot: true,                                                                  /* Include dependabot configuration. */
  // dependabotOptions: undefined,                                                      /* Options for dependabot. */
  // gitignore: undefined,                                                              /* Additional entries to .gitignore. */
  // jest: true,                                                                        /* Setup jest unit tests. */
  // jestOptions: undefined,                                                            /* Jest options. */
  // mergify: true,                                                                     /* Adds mergify configuration. */
  // mergifyAutoMergeLabel: 'auto-merge',                                               /* Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                         /* Options for mergify. */
  // npmignore: undefined,                                                              /* Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                            /* Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // projenDevDependency: true,                                                         /* Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                                 /* Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ '0 6 * * *' ],                                            /* Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                                    /* Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                                    /* Version of projen to install. */
  // pullRequestTemplate: true,                                                         /* Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                            /* The contents of the pull request template. */
  // rebuildBot: undefined,                                                             /* Installs a GitHub workflow which is triggered when the comment "@projen rebuild" is added to a pull request. */
  // rebuildBotCommand: 'rebuild',                                                      /* The pull request bot command to use in order to trigger a rebuild and commit of the contents of the branch. */
  releaseBranches: ['master'], /* Branches which trigger a release. */
  // releaseEveryCommit: true,                                                          /* Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                                        /* CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                               /* Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: undefined,                                                        /* Define a GitHub workflow for releasing from "master" when new versions are bumped. */
  // workflowBootstrapSteps: 'yarn install --frozen-lockfile && yarn projen',           /* Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                                 /* Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                                    /* The node version to use in GitHub workflows. */

  /* ProjectOptions */
  // clobber: true,                                                                     /* Add a `clobber` task which resets the repo to origin. */
  // devContainer: false,                                                               /* Add a VSCode development environment (used for GitHub Codespaces). */
  // gitpod: false,                                                                     /* Add a Gitpod development environment. */
  // logging: {},                                                                       /* Configure logging options such as verbosity. */
  // outdir: '.',                                                                       /* The root directory of the project. */
  // parent: undefined,                                                                 /* The parent project, if this project is part of a bigger project. */
  // projectType: ProjectType.UNKNOWN,                                                  /* Which type of project this is (library/app). */
  // readme: undefined,                                                                 /* The README setup. */
});

project.synth();


src/index.ts
import { Dashboard as cdkDashboard, DashboardProps, PeriodOverride } from '@aws-cdk/aws-cloudwatch';
import { Construct } from '@aws-cdk/core';
import { AutoScaling } from './auto-scaling';
import { LoadBalancer } from './load-balancer';
import { Rds } from './rds';
import { Redis } from './redis';

export interface HalloumiDashboard extends DashboardProps {
  /**
   * Name of the Load Balancer.
   *
   * If set, must only contain alphanumerics, dash (-) and underscore (_)
   *
   * @default - None
   * @stability stable
   */
  readonly loadBalancerName?: string;

  /**
   * Full Name of the Load Balancer.
   *
   * If set, must only contain alphanumerics, dash (-) and underscore (_)
   *
   * @default - None
   * @stability stable
   */
  readonly loadBalancerFullName?: string;

  /**
   * Name of the AutoScaling.
   *
   * If set, must only contain alphanumerics, dash (-) and underscore (_)
   *
   * @default - None
   * @stability stable
   */
  readonly autoScalingName?: string;

  /**
   * Max Capacity of the AutoScaling.
   *
   * If set, must only contain integer
   *
   * @default - 0
   * @stability stable
   */
  readonly autoScalingMaxCapacity?: number;

  /**
   * Name of the RDS.
   *
   * If set, must only contain alphanumerics, dash (-) and underscore (_)
   *
   * @default - None
   * @stability stable
   */
  readonly rdsName?: string;

  /**
   * Name of the Elasticache.
   *
   * If set, must only contain alphanumerics, dash (-) and underscore (_)
   *
   * @default - None
   * @stability stable
   */
  readonly elasticacheName?: string;
}

/**
 * An AWS CloudWatch Dashboard.
 *
 * @example
 * // create a dashboard for AutoScaling
 * new Dashboard(this, 'dashboard', {
 *    autoScalingName: 'my-auto-scaling',
 *    autoScalingMaxCapacity: 10
 * });
 *
 */
export class Dashboard extends Construct {
  /**
   * Creates a Dashboard based on the Halloumi best practices.
   *
   * @param scope the scope into which to import this dashboard
   * @param id the logical ID of the returned dashboard construct
   */
  constructor(scope: Construct, id: string, props?: HalloumiDashboard) {
    super(scope, id);

    let start = props?.start ? props.start : '-PT24H';
    let periodOverride = props?.periodOverride ? props.periodOverride : PeriodOverride.INHERIT;

    const dashboard = new cdkDashboard(this, id, {
      dashboardName: props?.dashboardName,
      start: start,
      end: props?.end,
      periodOverride: periodOverride,
      widgets: props?.widgets,
    });

    if (props?.loadBalancerFullName && props?.loadBalancerName) {
      const lbWidgets = LoadBalancer.metrics(props.loadBalancerName, props.loadBalancerFullName);
      lbWidgets.forEach(widget => {
        dashboard.addWidgets(widget);
      });
    }

    if (props?.autoScalingName && props.autoScalingMaxCapacity) {
      const autoScalingWidgets = AutoScaling.metrics(props.autoScalingName, props.autoScalingMaxCapacity);
      autoScalingWidgets.forEach(widget => {
        dashboard.addWidgets(widget);
      });
    }

    if (props?.rdsName) {
      const rdsWidgets = Rds.metrics(props.rdsName);
      rdsWidgets.forEach(widget => {
        dashboard.addWidgets(widget);
      });
    }

    if (props?.elasticacheName) {
      const redisWidgets = Redis.metrics(props.elasticacheName);
      redisWidgets.forEach(widget => {
        dashboard.addWidgets(widget);
      });
    }
  }
}

src/auto-scaling.ts
import { IWidget, TextWidget, Metric, GraphWidget, MathExpression } from '@aws-cdk/aws-cloudwatch';

export class AutoScaling {
  static metrics(name:string, max_capacity: number): IWidget[] {
    let header = new TextWidget({
      width: 24,
      height: 1,
      markdown: `# ${name} - Auto Scaling Group`,
    });
    let groupInServiceInstancesMetric = new Metric({
      namespace: 'AWS/AutoScaling',
      metricName: 'GroupInServiceInstances',
      dimensions: {
        AutoScalingGroupName: name,
      },
    });
    let GroupInServiceInstances = new GraphWidget({
      height: 6,
      width: 12,
      leftYAxis: {
        min: 0,
        label: 'Instances',
        showUnits: false,
      },
      leftAnnotations: [
        {
          label: 'Max Instances in ASG',
          value: max_capacity,
        },
      ],
    });
    GroupInServiceInstances.addRightMetric(groupInServiceInstancesMetric);
    GroupInServiceInstances.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: 'GroupInServiceInstances (expected)',
        color: '#666666',
        usingMetrics: {
          m1: groupInServiceInstancesMetric,
        },
      }),
    );

    let cpuUtilizationAverageMetric = new Metric({
      namespace: 'AWS/EC2',
      metricName: 'CPUUtilization',
      dimensions: {
        AutoScalingGroupName: name,
      },
      statistic: 'Average',
    });
    let cpuUtilizationAverage = new GraphWidget({
      height: 6,
      width: 12,
    });
    cpuUtilizationAverage.addRightMetric(cpuUtilizationAverageMetric);
    cpuUtilizationAverage.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: 'Average CPUUtilization (expected)',
        color: '#666666',
        usingMetrics: {
          m1: cpuUtilizationAverageMetric,
        },
      }),
    );

    let cpuUtilizationMaximumMetric = new Metric({
      namespace: 'AWS/EC2',
      metricName: 'CPUUtilization',
      dimensions: {
        AutoScalingGroupName: name,
      },
      statistic: 'Maximum',
    });
    let cpuUtilizationMaximum = new GraphWidget({
      height: 6,
      width: 12,
    });
    cpuUtilizationMaximum.addRightMetric(cpuUtilizationMaximumMetric);
    cpuUtilizationMaximum.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: 'Maximum CPUUtilization (expected)',
        color: '#666666',
        usingMetrics: {
          m1: cpuUtilizationMaximumMetric,
        },
      }),
    );

    let networkInMetric = new Metric({
      namespace: 'AWS/EC2',
      metricName: 'NetworkIn',
      dimensions: {
        AutoScalingGroupName: name,
      },
      statistic: 'Average',
    });
    let networkIn = new GraphWidget({
      height: 6,
      width: 12,
    });
    networkIn.addRightMetric(networkInMetric);
    networkIn.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: 'NetworkIn (expected)',
        color: '#666666',
        usingMetrics: {
          m1: networkInMetric,
        },
      }),
    );

    let networkOutMetric = new Metric({
      namespace: 'AWS/EC2',
      metricName: 'NetworkOut',
      dimensions: {
        AutoScalingGroupName: name,
      },
      statistic: 'Average',
    });
    let networkOut = new GraphWidget({
      height: 6,
      width: 12,
    });
    networkOut.addRightMetric(networkOutMetric);
    networkOut.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: 'NetworkOut (expected)',
        color: '#666666',
        usingMetrics: {
          m1: networkOutMetric,
        },
      }),
    );

    return [header, GroupInServiceInstances, cpuUtilizationAverage, cpuUtilizationMaximum, networkIn, networkOut];
  }
}

src/redis.ts
import { TextWidget, MathExpression, GraphWidget, Metric, IWidget } from '@aws-cdk/aws-cloudwatch';

export class Redis {

  static metrics(name:string): IWidget[] {
    let header = new TextWidget({
      width: 24,
      height: 1,
      markdown: `# ${name} - Redis`,
    });

    let metrics = [
      'CPUUtilization',
      'FreeableMemory',
      'SwapUsage',
      'NetworkBytesIn',
      'NetworkBytesOut',
      'CurrConnections',
      'Evictions',
      'Reclaimed',
      'CacheHits',
      'CacheMisses',
      'ReplicationBytes',
      'ReplicationLag',
      'BytesUsedForCache',
    ];

    let graph = Array();
    graph.push(header);

    metrics.forEach(metric => {
      graph.push(
        Redis.metricGraph(metric, name),
      );
    });
    return graph;
  }

  static metricGraph(metric: string, name: string): IWidget {
    let cloudwatchMetric = new Metric({
      namespace: 'AWS/ElastiCache',
      metricName: metric,
      dimensions: {
        CacheClusterId: name,
      },
    });
    let graphWidget = new GraphWidget({
      height: 6,
      width: 12,
    });
    graphWidget.addLeftMetric(
      new MathExpression({
        expression: 'ANOMALY_DETECTION_BAND(m1, 2)',
        label: `${metric} (expected)`,
        color: '#666666',
        usingMetrics: {
          m1: cloudwatchMetric,
        },
      }),
    );
    graphWidget.addRightMetric(cloudwatchMetric);
    return graphWidget;
  }
}